### PR TITLE
[type:feat] shenyu client heartbeat

### PR DESCRIPTION
**Pull request by Aias00**
add shenyu client heartbeat

Make sure that:

- [x] You have read the [contribution guidelines](https://shenyu.apache.org/community/contributor-guide).
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] Your local test passed `./mvnw clean install -Dmaven.javadoc.skip=true`.

### CODE REVIEW

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/AbstractShenyuClientRegisterServiceImpl.java [hunk 1]
```diff
@@ -196,7 +200,40 @@ public String doRegisterURI(final String selectorName, final List<URIRegisterDTO
         }
         return ShenyuResultMessage.SUCCESS;
     }
-
+    
+    @Override
+    public String doHeartbeat(final String selectorName, final List<URIRegisterDTO> uriList) {
+        if (CollectionUtils.isEmpty(uriList)) {
+            return "";
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/AbstractShenyuClientRegisterServiceImpl.java:207]: return null?

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/AbstractShenyuClientRegisterServiceImpl.java (hunk 1 above)
Comment (Aias00) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/AbstractShenyuClientRegisterServiceImpl.java:207]: i copied this from 'org.apache.shenyu.admin.service.register.AbstractShenyuClientRegisterServiceImpl#doRegisterURI'

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/AbstractShenyuClientRegisterServiceImpl.java [hunk 2]
```diff
@@ -196,7 +200,43 @@ public String doRegisterURI(final String selectorName, final List<URIRegisterDTO
...
+        }
+        String pluginName = PluginNameAdapter.rpcTypeAdapter(rpcType());
+        // todo:[To be refactored with namespace] Temporarily hardcode
+        SelectorDO selectorDO = selectorService.findByNameAndPluginNameAndNamespaceId(selectorName, pluginName, SYS_DEFAULT_NAMESPACE_ID);
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/AbstractShenyuClientRegisterServiceImpl.java:211]: use request namespaceId

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/AbstractShenyuClientRegisterServiceImpl.java [hunk 3]
```diff
@@ -196,7 +200,43 @@ public String doRegisterURI(final String selectorName, final List<URIRegisterDTO
...
+        validUriList.forEach(uriRegisterDTO -> {
+            DiscoveryUpstreamDTO discoveryUpstreamDTO = CommonUpstreamUtils.buildDefaultDiscoveryUpstreamDTO(uriRegisterDTO.getHost(),
+                    uriRegisterDTO.getPort(),
+                    uriRegisterDTO.getProtocol(),
+                    uriRegisterDTO.getNamespaceId());
+            LOG.info("change alive selectorId={}|url={}", selectorId, discoveryUpstreamDTO.getUrl());
+            discoveryUpstreamService.changeStatusBySelectorIdAndUrl(selectorId, discoveryUpstreamDTO.getUrl(), Boolean.TRUE);
+        });
+        
+        DiscoverySyncData discoverySyncData = fetch(selectorId, selectorDO.getName(), pluginName, SYS_DEFAULT_NAMESPACE_ID);
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/AbstractShenyuClientRegisterServiceImpl.java:234]: same question

File: shenyu-admin/src/test/java/org/apache/shenyu/admin/service/register/FallbackShenyuClientRegisterServiceTest.java [hunk 4]
```diff
@@ -72,7 +77,12 @@ static class MockFallbackShenyuClientRegisterServiceException extends FallbackSh
         String doRegisterURI(final String selectorName, final List<URIRegisterDTO> uriList) {
             throw new ShenyuException("Exception");
         }
-
+        
+        @Override
+        String doHeartbeat(final String selectorName, final List<URIRegisterDTO> uriList) {
+            throw new ShenyuException("Exception");
```
Comment (moremind) [inline shenyu-admin/src/test/java/org/apache/shenyu/admin/service/register/FallbackShenyuClientRegisterServiceTest.java:83]: why exception?

Comment (Aias00) [reply shenyu-admin/src/test/java/org/apache/shenyu/admin/service/register/FallbackShenyuClientRegisterServiceTest.java:83]: this is a Exception class, for Exception test, copied from org.apache.shenyu.admin.service.register.FallbackShenyuClientRegisterServiceTest.MockFallbackShenyuClientRegisterServiceException#doRegisterURI

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/ShenyuClientRegisterDivideServiceImpl.java [hunk 5]
```diff
@@ -86,6 +86,10 @@ protected String buildHandle(final List<URIRegisterDTO> uriList, final SelectorD
         if (isEventDeleted) {
             addList.get(0).setStatus(false);
         }
+        boolean isEventHeartbeat = uriList.size() == 1 && EventType.HEARTBEAT.equals(uriList.get(0).getEventType());
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/ShenyuClientRegisterDivideServiceImpl.java:89]: why urlList == 1?not urlList > 0?

**Review by moremind — CHANGES_REQUESTED** [review]

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/ShenyuClientRegisterDivideServiceImpl.java (hunk 5 above)
Comment (Aias00) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/ShenyuClientRegisterDivideServiceImpl.java]: i copied from other place, changed

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/ShenyuClientRegisterDivideServiceImpl.java [hunk 6]
```diff
@@ -82,10 +82,14 @@ protected void registerMetadata(final MetaDataRegisterDTO dto) {
     protected String buildHandle(final List<URIRegisterDTO> uriList, final SelectorDO selectorDO) {
         List<DivideUpstream> addList = buildDivideUpstreamList(uriList);
         List<DivideUpstream> canAddList = new CopyOnWriteArrayList<>();
-        boolean isEventDeleted = uriList.size() == 1 && EventType.DELETED.equals(uriList.get(0).getEventType());
+        boolean isEventDeleted = !uriList.isEmpty() && EventType.DELETED.equals(uriList.get(0).getEventType());
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/ShenyuClientRegisterDivideServiceImpl.java:85]: CollectionUtils.isNotEmpty

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/ShenyuClientRegisterDivideServiceImpl.java [hunk 7]
```diff
@@ -82,10 +82,14 @@ protected void registerMetadata(final MetaDataRegisterDTO dto) {
...
         if (isEventDeleted) {
             addList.get(0).setStatus(false);
         }
+        boolean isEventHeartbeat = !uriList.isEmpty() && EventType.HEARTBEAT.equals(uriList.get(0).getEventType());
+        if (isEventHeartbeat) {
+            addList.get(0).setStatus(true);
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/service/register/ShenyuClientRegisterDivideServiceImpl.java:91]: why get(0)

**Review by moremind — CHANGES_REQUESTED** [review]

**Review by moremind — APPROVED** [review]

**Pull request merged**